lua/swampshop/sh_products.lua

```diff
@@ -78,33 +78,55 @@ function SS_WeaponProduct(product)
     product.SS_WeaponProduct_OnBuy = product.OnBuy or function() end
 
     function product:OnBuy(ply)
-        ply:Give(self.class)
-        ply:SelectWeapon(self.class)
-        self:SS_WeaponProduct_OnBuy(ply)
+        if(!ply:HasWeapon(self.Class) )then
+            ply:Give(self.class)
+            self:SS_WeaponProduct_OnBuy(ply)
+        else
+            ply:SelectWeapon(self.class)
+        end
+    end
+    function product:CannotBuy(ply)
+        return !ply:Alive() and "You're dead!"
     end
-
     SS_DeathKeepnotice(product)
     SS_Product(product)
 end
 
 function SS_WeaponAndAmmoProduct(product)
     function product:OnBuy(ply)
         if not ply:HasWeapon(self.class) then
-            ply:Give(self.class)
+            local wep = ply:Give(self.class)
+            
+            if(self.amount)then --ensure the player has at least this many ammo when giving a new one to them. Otherwise, give them whatever is standard.
+                local ammotype = self.ammotype or game.GetAmmoName(wep:GetPrimaryAmmoType())
+                local curammo = ply:GetAmmoCount(self.ammotype)
+                ply:SetAmmo(math.max(curammo,self.amount),ammotype)
+            end
+        else
             local wep = ply:GetWeapon(self.class)
+            local ammotype = self.ammotype or game.GetAmmoName(wep:GetPrimaryAmmoType())
+            local ammogive = self.amount or (wep.Primary and wep.Primary.DefaultClip) or wep:GetMaxClip1() or 0
 
-            if wep:Clip1() > 0 then
-                wep:SetClip1(0)
+            if (ammotype and ammogive > 0) then
+                ply:GiveAmmo(ammogive, ammotype)
             end
-
-            ply:SetAmmo(self.amount, self.ammotype)
-        else
-            ply:GiveAmmo(self.amount, self.ammotype)
         end
-
         ply:SelectWeapon(self.class)
     end
-
+    function product:CannotBuy(ply)
+        if !ply:Alive() then return "You're dead!" end
+        if (ply:HasWeapon(self.class)) then
+            local wep = ply:GetWeapon(self.class)
+            local ammotype = self.ammotype or (game.GetAmmoName(wep:GetPrimaryAmmoType()))
+            local ammogive = self.amount or (wep.Primary and wep.Primary.DefaultClip) or wep:GetMaxClip1() or 0
+            local fail = assert(ammotype != nil and ammogive > 0,"shop purchase could not refill" .. ply:Nick() .. "'s "..self.class.." with ".. ammogive .." of "..(ammotype or "<nil ammotype>") ,true)
```

Comment:
This isn't what assert does...? https://wiki.facepunch.com/gmod/Global.assert

Reply:
I looked at this and I don't entirely understand what it returns.